# Run All

In [ ]:
"""
Prazo:
24/10/2022 (8 dias)

Slide 1 - Abertura 80/20: Explicar o motivo da decisão e Inicar a apresentacação falando do 80/20.
Slide 2 - Resumo Estatístico: Em seguida apresentar um breve dashboard sobre o resumo estatístico e financeiro da empresa nesses 3 anos. (https://app.diagrams.net/#G1DPfE2taGdSphMyXd9k1pYCJ2WxLulLSn)
          2.1) Soma Total de Vendas (Brutas e Liquidas) Por Ano. 
          2.2) Média Total de Vendas Por Ano
          2.3) Média de (Média de Quantidade Liquida por Grupo)
          2.4) Média de (Média de Total de Pedidos por Mes)
          2.5) Análise Sazonal das vendas líquidas por Mes
          2.6) Análise de Tendência das vendas líquidas por Mes
Slide 3 - Perguntas de Negócio: Mostrar em 1 slide com Storyteling:
          3.1) Quais grupos são mais/menos vantajogos financeiramente (Lucro Líquido)?
          3.2) Perguntas e Tabela Exemplo:
                3.2.1) Top 5 grupos que mais/menos oferece desconto e quais as médias (A média é uma boa estatística?). 
                3.2.2) Qual o percentual de vendas com desconto? 
                3xz.2.3) Qual o percentual de vendas com devoluções (Entre parenteses ao lado: Qual o valor médio dessas devoluções?
                [Grupos | Top(±) 5 Media_Desconto | Percentual_Vendas_Com_Desconto | Percentual_Vendas_Devolvidas]
          3.3) Meu desconto oferecido está adequado ou seria indicado realizar alterações? Qual a tendência? (Séries Temporais)
          3.4) Qual a tendência de desconto? (para cada grupo e para a empresa) (Séries Temporais)
Slide 4 - Correlações: 
          4.1) Qual a relação (por grupo) entre quantidade_liquida e custo_produto e o que podemos aprender com isso? (Heatmap)
          4.2) Qual a relação (por grupo) entre desconto e devolução e o que podemos aprender com isso? (Heatmap)
Slide 5 - Análise Mensal: Qual é o valor médio para cada mês: Vendas, Descontos e Devoluções. (Gráfico de múltiplas colunas verticais)
Slide 6 - Oportunidade de melhoria:
          6.1) Foco no 80/20.
          6.2) Melhoria no Desconto.
          6.3) Cuidado com Produtos mais devolvidos (entender melhor o motivo e diminuir as ocorrências).
"""

'\nPrazo:\n24/10/2022 (8 dias)\n\nSlide 1 - Abertura 80/20: Explicar o motivo da decisão e Inicar a apresentacação falando do 80/20.\nSlide 2 - Resumo Estatístico: Em seguida apresentar um breve dashboard sobre o resumo estatístico e financeiro da empresa nesses 3 anos. (https://app.diagrams.net/#G1DPfE2taGdSphMyXd9k1pYCJ2WxLulLSn)\n          2.1) Soma Total de Vendas (Brutas e Liquidas) Por Ano. \n          2.2) Média Total de Vendas Por Ano\n          2.3) Média de (Média de Quantidade Liquida por Grupo)\n          2.4) Média de (Média de Total de Pedidos por Mes)\n          2.5) Análise Sazonal das vendas líquidas por Mes\n          2.6) Análise de Tendência das vendas líquidas por Mes\nSlide 3 - Perguntas de Negócio: Mostrar em 1 slide com Storyteling:\n          3.1) Quais grupos são mais/menos vantajogos financeiramente (Lucro Líquido)?\n          3.2) Perguntas e Tabela Exemplo:\n                3.2.1) Top 5 grupos que mais/menos oferece desconto e quais as médias (A média é uma

In [ ]:
import pandas as pd
import numpy as np
import plotly.express as px

In [ ]:
ordem_vendas = pd.read_csv('/content/drive/MyDrive/Uelliton/vendas_online_2017_2019/dados_originais/business.retailsales.csv')
vendas_anuais = pd.read_csv('/content/drive/MyDrive/Uelliton/vendas_online_2017_2019/dados_originais/business.retailsales2.csv')
calendario = pd.read_csv('/content/drive/MyDrive/Uelliton/vendas_online_2017_2019/dados_tratados/calendario.csv')

In [ ]:
calendario['data_num'] = pd.to_datetime(calendario['data_num'])
calendario.sort_values(by='data_num',inplace=True)
calendario.reset_index(inplace=True)

In [ ]:
ordem_vendas.columns = ['grupo_do_produto', 'quantidade_liquida', 'venda_bruta', 'custo_produto', 'retorno', 'total_vendas_liquida']
ordem_vendas.head()

grupo_do_produto  quantidade_liquida  venda_bruta  custo_produto  retorno  \
0  Art & Sculpture                  34      14935.0        -594.00  -1609.0   
1           Basket                  13       3744.0        -316.80      0.0   
2           Basket                  12       3825.0        -201.60   -288.0   
3           Basket                  17       3035.0         -63.25      0.0   
4  Art & Sculpture                  47       2696.8         -44.16      0.0   

   total_vendas_liquida  
0              12732.00  
1               3427.20  
2               3335.40  
3               2971.75  
4               2652.64

In [ ]:
ordem_vendas.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1775 entries, 0 to 1774
Data columns (total 6 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   grupo_do_produto      1767 non-null   object 
 1   quantidade_liquida    1775 non-null   int64  
 2   venda_bruta           1775 non-null   float64
 3   custo_produto         1775 non-null   float64
 4   retorno               1775 non-null   float64
 5   total_vendas_liquida  1775 non-null   float64
dtypes: float64(4), int64(1), object(1)
memory usage: 83.3+ KB


In [ ]:
ordem_vendas.sample(5)

grupo_do_produto  quantidade_liquida  venda_bruta  custo_produto  \
899             Basket                   1         94.0           0.00   
775             Basket                   1        115.0           0.00   
455            Jewelry                   7        196.0         -16.79   
1345      Recycled Art                   1         44.0           0.00   
213   Fair Trade Gifts                  16        352.0         -13.18   

      retorno  total_vendas_liquida  
899       0.0                 94.00  
775       0.0                115.00  
455       0.0                179.21  
1345      0.0                 44.00  
213       0.0                338.82

In [ ]:
lista_cols = ['mes','ano','total_pedidos', 'venda_bruta', 'desconto', 'custo_produto', 'venda_liquida', 'transporte', 'total_vendas']
if len(lista_cols) == len(vendas_anuais.columns):
  vendas_anuais.columns = lista_cols
vendas_anuais['data'] = calendario['data_num']
vendas_anuais['%_desconto'] = round(((-1* vendas_anuais['desconto'])/vendas_anuais['venda_bruta'])*100,2)
vendas_anuais.head()

mes   ano  total_pedidos  venda_bruta  desconto  custo_produto  \
0   January  2017             73       8861.5   -129.40        -448.45   
1  February  2017             56       6908.5   -104.70        -416.20   
2     March  2017             60       5778.5   -172.20       -1017.20   
3     April  2017             70       8814.0   -281.40           0.00   
4       May  2017             54       6677.0   -185.75        -253.80   

   venda_liquida  transporte  total_vendas       data  %_desconto  
0        8283.65     1088.30       9371.95 2017-01-01        1.46  
1        6387.60      892.45       7280.05 2017-01-02        1.52  
2        4589.10      707.43       5296.53 2017-01-03        2.98  
3        8532.60     1068.30       9600.90 2017-01-04        3.19  
4        6237.45      866.46       7103.91 2017-01-05        2.78

In [ ]:
# # Remoção de Outliers
# # Vou remover os outliers superiores de 2019 pois não condizem com a realidade dos dados. Análises futuras podem analisar os casos individualmente para aprender sobre eles.
# vendas_anuais.drop(vendas_anuais[vendas_anuais.total_pedidos == 342].index, inplace=True)
# vendas_anuais.drop(vendas_anuais[vendas_anuais.total_pedidos == 272].index, inplace=True)
# # Vou remover o outlier inferior em 2017 pois não condiz com a realidade dos dados. Análises futuras podem analisar o caso individualmente para aprender sobre ele.
# vendas_anuais.drop(vendas_anuais[vendas_anuais.custo_produto == -1017.2].index, inplace=True)

In [ ]:
vendas_anuais.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 36 entries, 0 to 35
Data columns (total 11 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   mes            36 non-null     object        
 1   ano            36 non-null     int64         
 2   total_pedidos  36 non-null     int64         
 3   venda_bruta    36 non-null     float64       
 4   desconto       36 non-null     float64       
 5   custo_produto  36 non-null     float64       
 6   venda_liquida  36 non-null     float64       
 7   transporte     36 non-null     float64       
 8   total_vendas   36 non-null     float64       
 9   data           36 non-null     datetime64[ns]
 10  %_desconto     36 non-null     float64       
dtypes: datetime64[ns](1), float64(7), int64(2), object(1)
memory usage: 3.2+ KB


In [ ]:
vendas_anuais.sample(5)

mes   ano  total_pedidos  venda_bruta  desconto  custo_produto  \
10   November  2017             91     13025.00   -131.30        -323.85   
20  September  2018             79     10582.85   -276.15           0.00   
33    October  2019             97      9132.25   -279.42       -1572.55   
19     August  2018             82      7698.70   -140.57        -197.35   
17       June  2018            124     13260.80   -335.40       -1506.53   

    venda_liquida  transporte  total_vendas       data  %_desconto  
10       12569.85     1555.10      14124.95 2017-01-11        1.01  
20       10306.70     1634.33      11941.03 2018-01-09        2.61  
33        7280.28     1631.25       8911.53 2019-01-10        3.06  
19        7360.78     1404.03       8764.81 2018-01-08        1.83  
17       11418.87     2124.49      13543.36 2018-01-06        2.53



---



## Análise 80/20

In [ ]:
# Quais tipos de produtos contribuem com 80% das vendas

# Somatória do total de vendas liquidas por grupo
pivot_1 = ordem_vendas.groupby(["grupo_do_produto"])[["total_vendas_liquida"]].agg("sum")
# A divisão entre o total de vendas de cada grupo pela somatoria de todos os grupos (proporção)
pivot_1["proporcao_vendas"] = (pivot_1["total_vendas_liquida"]/pivot_1["total_vendas_liquida"].sum())
# multiplicando a porporcao *100 para pegar a porcentagem
pivot_1["porcentagem_vendas"] = round( (pivot_1["proporcao_vendas"])*100, 2)
# Ordenando os valores de proporcao de vendas para decrescente (os maiores acima)
pivot_1.sort_values("proporcao_vendas", ascending = False, inplace = True)
# Aplicando uma soma cumulativa na coluna proporcao de vendas. Aplicando uma funcao anonima para arredondar os valores para duas casas decimais e multiplicar *100
pivot_1["proporcao_vendas_acumulada"] = pivot_1["proporcao_vendas"].cumsum().apply(lambda x: "%.2f" % round(100*x,2))

# A media do total de vendas liquidas agrupado por grupo
pivot_2 = ordem_vendas.groupby(["grupo_do_produto"])[["total_vendas_liquida"]].agg("mean")
# Arredondando os valores para duas casas decimais
pivot_2["media_vendas_por_grupo"] = pivot_2["total_vendas_liquida"].apply(lambda x: "%.2f" % round(x,2))
# Renomeando a coluna para um nome que melhor descreve os dados
pivot_2.rename(columns = {"total_vendas_liquida":"media_vendas_por_grupo"}, inplace = True)

# Contagem de vendas agrupado por grupos
pivot_3 = ordem_vendas.groupby(["grupo_do_produto"])[["total_vendas_liquida"]].agg("count")
# Renomeando a coluna para um nome que melhor descreve os dados
pivot_3.rename(columns = {"total_vendas_liquida":"soma_total_de_pedidos"}, inplace = True)

# Realizando as mesclagens
pivot_1 = pivot_1.join(pivot_2)
pivot_1 = pivot_1.join(pivot_3)

In [ ]:
# Exportando a tabela para CSV para acessar pelo PowerBI
pivot_1.to_csv('/content/drive/MyDrive/Uelliton/vendas_online_2017_2019/dados_tratados/80_20.csv')

In [ ]:
# Os 4 grupos que representam 80% do faturamento
pivot_1.head(4)

total_vendas_liquida  proporcao_vendas  porcentagem_vendas  \
grupo_do_produto                                                               
Basket                       134791.39          0.404677               40.47   
Art & Sculpture               84480.85          0.253632               25.36   
Jewelry                       29572.95          0.088785                8.88   
Home Decor                    25699.99          0.077158                7.72   

                 proporcao_vendas_acumulada media_vendas_por_grupo  \
grupo_do_produto                                                     
Basket                                40.47             244.630472   
Art & Sculpture                       65.83             250.685015   
Jewelry                               74.71             140.823571   
Home Decor                            82.43             196.183130   

                 media_vendas_por_grupo  soma_total_de_pedidos  
grupo_do_produto                                                
Basket                           244.63                    551  
Art & Sculpture                  250.69                    337  
Jewelry                          140.82                    210  
Home Decor                       196.18                    131

In [ ]:
# Top 10 Grupos que MENOS contribui para minhas vendas 
pivot_1.tail(10)

total_vendas_liquida  proporcao_vendas  porcentagem_vendas  \
grupo_do_produto                                                               
Kids                           3721.34          0.011172                1.12   
Recycled Art                   3704.16          0.011121                1.11   
Skin Care                      2571.80          0.007721                0.77   
Music                          2418.90          0.007262                0.73   
Fair Trade Gifts               2204.67          0.006619                0.66   
One-of-a-Kind                  2108.01          0.006329                0.63   
Furniture                      1864.96          0.005599                0.56   
Textiles                       1679.10          0.005041                0.50   
Easter                           34.20          0.000103                0.01   
Gift Baskets                     19.50          0.000059                0.01   

                 proporcao_vendas_acumulada media_vendas_por_grupo  \
grupo_do_produto                                                     
Kids                                  95.01              59.068889   
Recycled Art                          96.13             161.050435   
Skin Care                             96.90             233.800000   
Music                                 97.63              83.410345   
Fair Trade Gifts                      98.29              78.738214   
One-of-a-Kind                         98.92             175.667500   
Furniture                             99.48             116.560000   
Textiles                              99.98             119.935714   
Easter                                99.99              34.200000   
Gift Baskets                         100.00              19.500000   

                 media_vendas_por_grupo  soma_total_de_pedidos  
grupo_do_produto                                                
Kids                              59.07                     63  
Recycled Art                     161.05                     23  
Skin Care                        233.80                     11  
Music                             83.41                     29  
Fair Trade Gifts                  78.74                     28  
One-of-a-Kind                    175.67                     12  
Furniture                        116.56                     16  
Textiles                         119.94                     14  
Easter                            34.20                      1  
Gift Baskets                      19.50                      1

Basket, Art & Sculpture	(Arte e Escultura), Jewelry (Jóias) e Home Decor (Decoração) são os *4 principais grupos de produtos* (de 18) que contribuem com *mais de 82.43% do Total de suas Vendas Líquidas*.

Quando observamos os *Grupos de Produtos que possuem mais vendas*, esses 4 grupos de produtos aparecem novamente.

No entanto, não se sabe se esses 4 grupos de produtos são os mesmos para cada um dos 3 anos. Além disso, com a falta de dados, não é possível determinar se esses 4 principais grupos se devem ao aumento das vendas em novembro e dezembro. Para isso, é necessário o timestamp dos pedidos para aprimorar a análise.

Se esses 4 grupos de produtos forem os mesmos durante todo o ano, a(o) cliente deve focar mais esforços na geração de mais pedidos e vendas para esses produtos.

Caso cortar gastos diminuindo a gama de produtos e focar em grupos mais rentáveis, deve-se dar preferência aos grupos Top -10 que correspondem apenas a 5% de todo o lucro.

## Análise Exploratória/Estatística

Slide 2
    
    Resumo Estatístico: Em seguida apresentar um breve dashboard sobre o resumo estatístico e financeiro da empresa nesses 3 anos. 

Diagrama:

https://app.diagrams.net/#G1DPfE2taGdSphMyXd9k1pYCJ2WxLulLSn

O que responder:

    2.1) Soma Total de Vendas (Brutas e Liquidas) Por Ano. 
    2.2) Média Total de Vendas Por Ano
    2.3) Média de (Média de Quantidade Liquida por Grupo)
    2.4) Média de (Média de Total de Pedidos por Mes)
    2.5) Análise Sazonal das vendas líquidas por Mes
    2.6) Análise de Tendência das vendas líquidas por Mes

In [ ]:
# 2.1) Soma Total de Vendas (Brutas e Liquidas) Por Ano. 
# 2.2) Média Total de Vendas Por Ano
vendas_por_ano = vendas_anuais.groupby('ano').agg(['sum','mean'])[['venda_bruta','venda_liquida']]
# vendas_por_ano.to_csv('/content/drive/MyDrive/Uelliton/vendas_online_2017_2019/dados_tratados/vendas_por_ano.csv')
vendas_por_ano

venda_bruta               venda_liquida              
             sum          mean           sum          mean
ano                                                       
2017    97342.05   8111.837500      91520.50   7626.708333
2018   113313.15   9442.762500     104788.18   8732.348333
2019   143762.15  11980.179167     129796.40  10816.366667

In [ ]:
# Soma em todos os anos
vendas_por_grupo = ordem_vendas.groupby(["grupo_do_produto"])[["total_vendas_liquida"]].agg("sum")
# A divisão entre o total de vendas de cada grupo pela somatoria de todos os grupos (proporção)
vendas_por_grupo["%_vendas"] = round( (vendas_por_grupo["total_vendas_liquida"]/vendas_por_grupo["total_vendas_liquida"].sum())*100,2)
# vendas_por_grupo.to_csv('/content/drive/MyDrive/Uelliton/vendas_online_2017_2019/dados_tratados/vendas_por_grupo.csv')
vendas_por_grupo.head(3)

total_vendas_liquida  %_vendas
grupo_do_produto                                
Accessories                    3785.38      1.14
Art & Sculpture               84480.85     25.36
Basket                       134791.39     40.47

In [ ]:
# 2.4) Análise de Tendencia será tratada direto no powerbi (ou nao)
analise_temporal = vendas_anuais.copy()
# analise_temporal.to_csv('/content/drive/MyDrive/Uelliton/vendas_online_2017_2019/dados_tratados/analise_temporal.csv')

In [ ]:
#analise_temporal['desconto'] = analise_temporal['desconto']*(-1)
#analise_temporal['custo_produto'] = analise_temporal['custo_produto']*(-1)
analise_temporal.drop('data',axis=1).style.background_gradient()



---



### Análise de Série Temporal

In [ ]:
# colunas: venda_bruta, venda_liquida, perc_desconto, total_vendas, transporte, custo_produto, desconto 
# https://pt.planetcalc.com/7898/?_d=dTKQFtAdiu.a1le-7wGh5YKuWT5Bvv-GIXfx-KJRCVs2RwIRjIYT6C6d.6Y_

In [ ]:
# É série estacionaria?
# Qual a tendencia?
# Tendencia Aditiva ou Multiplicativa?
# Qual o periodo das sazonalidades
# Tendencia de longo prazo? 6 meses +-


## **Análise Anual**

In [ ]:
# Vou remover os outliers superiores de 2019 pois não condizem com a realidade dos dados. Análises futuras podem analisar os casos individualmente para aprender sobre eles.
px.box(vendas_anuais, y='total_pedidos',facet_col='ano', points='all')

In [ ]:
px.box(vendas_anuais, y='venda_liquida',facet_col='ano', points='all')

In [ ]:
# Vou remover o outlier inferior em 2017 pois não condiz com a realidade dos dados. Análises futuras podem analisar o caso individualmente para aprender sobre ele.
px.box(vendas_anuais, y='custo_produto',facet_col='ano', points='all')

In [ ]:
px.box(vendas_anuais, y='%_desconto',facet_col='ano', points='all')



---



### Análise de Série Temporal

In [ ]:
temporal = vendas_anuais.copy()
# tem tendencia na serie?R: usar correlacao de spearman para mensurar. Caso o coef. for negativo entao é 
# uma serie estacionaria ou seja, nao tem dentencia ou positivo para nao estacionaria com tendencia

# caso seja estacionaria: Análise de média movel de 2 ou 3 meses para diminuir o impacto das variacoes e estudo da tendencia


In [ ]:
temporal

mes   ano  total_pedidos  venda_bruta  desconto  custo_produto  \
0     January  2017             73      8861.50   -129.40        -448.45   
1    February  2017             56      6908.50   -104.70        -416.20   
2       March  2017             60      5778.50   -172.20       -1017.20   
3       April  2017             70      8814.00   -281.40           0.00   
4         May  2017             54      6677.00   -185.75        -253.80   
5        June  2017             68      9621.50   -234.45         -17.50   
6        July  2017             66      6480.00    -51.50        -469.20   
7      August  2017             55      8025.00   -258.90         -26.00   
8   September  2017             68      7075.00    -61.70        -281.00   
9     October  2017             59      5720.00    -88.00        -305.00   
10   November  2017             91     13025.00   -131.30        -323.85   
11   December  2017            116     10356.05   -149.85        -414.20   
12    January  2018             83      8923.00   -217.10         -26.25   
13   February  2018             69      6529.20   -161.35        -118.15   
14      March  2018             64      7442.70   -226.82          -8.80   
15      April  2018             81      9406.35   -232.28         -40.00   
16        May  2018             82      7493.90   -221.25       -1448.02   
17       June  2018            124     13260.80   -335.40       -1506.53   
18       July  2018            102      9274.95   -237.87        -689.98   
19     August  2018             82      7698.70   -140.57        -197.35   
20  September  2018             79     10582.85   -276.15           0.00   
21    October  2018             71      7014.50   -277.95        -294.75   
22   November  2018            140     12002.70   -414.45        -154.40   
23   December  2018            164     13683.50   -371.20        -928.35   
24    January  2019             87      7811.75   -261.97       -1250.35   
25   February  2019             63      6523.70   -288.70         -38.70   
26      March  2019             99     13075.50   -439.85        -981.00   
27      April  2019             92      6977.00   -285.40         -10.00   
28        May  2019             96     10082.25   -460.90        -173.35   
29       June  2019             85      7321.75   -186.02        -165.40   
30       July  2019             94      9739.20   -447.07        -908.75   
31     August  2019            105      8839.50   -201.67         -84.37   
32  September  2019             88      9077.45   -354.89        -853.35   
33    October  2019             97      9132.25   -279.42       -1572.55   
34   November  2019            272     23997.90   -776.84        -364.51   
35   December  2019            342     31183.90  -2269.51       -1311.18   

    venda_liquida  transporte  total_vendas       data  %_desconto  
0         8283.65     1088.30       9371.95 2017-01-01        1.46  
1         6387.60      892.45       7280.05 2017-01-02        1.52  
2         4589.10      707.43       5296.53 2017-01-03        2.98  
3         8532.60     1068.30       9600.90 2017-01-04        3.19  
4         6237.45      866.46       7103.91 2017-01-05        2.78  
5         9369.55     1204.32      10573.87 2017-01-06        2.44  
6         5959.30      807.36       6766.66 2017-01-07        0.79  
7         7740.10      843.46       8583.56 2017-01-08        3.23  
8         6732.30      907.32       7639.62 2017-01-09        0.87  
9         5327.00      695.42       6022.42 2017-01-10        1.54  
10       12569.85     1555.10      14124.95 2017-01-11        1.01  
11        9792.00     1340.85      11132.85 2017-01-12        1.45  
12        8679.65     1180.18       9859.83 2018-01-01        2.43  
13        6249.70      908.91       7158.61 2018-01-02        2.47  
14        7207.08     1226.92       8434.00 2018-01-03        3.05  
15        9134.07     1387.56      10521.63 2018-01-04        2.47  
16        5824.63     12

In [ ]:
temporal['data'].dt.day

0      1
1      2
2      3
3      4
4      5
5      6
6      7
7      8
8      9
9     10
10    11
11    12
12     1
13     2
14     3
15     4
16     5
17     6
18     7
19     8
20     9
21    10
22    11
23    12
24     1
25     2
26     3
27     4
28     5
29     6
30     7
31     8
32     9
33    10
34    11
35    12
Name: data, dtype: int64